addons/website/tests/test_views.py

```diff
@@ -300,6 +300,52 @@ def setUp(self):
             'key': 'website.extension_view',
         })
 
+    def test_module_new_inherit_view_with_grandparent_specific(self):
```

Comment:
This test is doing something weird, I am trying to figure.
Let's call the views A B C D and A' B' C' D' as we generally do, where `'` means it is a COW view.

It does:
1. Create A (g4) (in the backend) 
2. COW A
3. Create B (g3) (in the backend)
4. Create C (g2) (in the backend)
5. COW C (ge)
6. Create D (g1) (through module install)

But step 3. is never supposed to happen, unless people are playing in the backend in debug mode in views, and we don't handle such cases as it would simply be impossible to cover those flows on top of the current COW mechanism handled by the `write/unlink/create` + module install methods. 

So at step 3, you have
```
A      A'
 |
B
```
Which can not happen with:
1. **COW mechanism** as it would have created a B' when cowing A (if B was there before A') or would have created a B' bellow A' when creating B (if A' was there before B)
2. **Module install** as it would have created a B' under A' if B was created during module install and A' existed (the other case from point 1. is not possible through module install, no reason to create A')

Reply:
This is the simplest case I could reproduce from a failing upgrade that fails with the same error. 
I have no knowledge about valid flows.

The case where this is reproduced is on an upgrade 13.0->14.0, not just a module upgrade.
This means that multiple views are created, removed, and updated from multiple modules.
The views involved in the failing upgrade are from portal, web, website, and website_sale. There are also several theme views updated/removed.
 
Another weird thing (not sure if valid flow) I noticed on the upgrading DB is that there is a specific view, lets call it X, with a generic child without specific counterpart. X is pointed by a standard xmlid and during the upgrade it is assigned NULL inherit_id (making it generic). In the end this was not the direct cause of this issue but it triggered some extra COWs.

When the case above happens, and since the SQL query does not check for `generic_parent.website_id IS NULL`, then `generic_parent` ends up being also `specific_parent`. If I understand correctly what was done via ORM, and was replaced by the SQL query, actually checked that `generic_parent.website_id IS NULL` (`not record.inherit_id.website_id`). See https://github.com/odoo/odoo/commit/56f6e1a6b8d1eee8b8f8556ac6c24d947b44f19e#diff-55a5ba4d6926898a5b68632527b57bc496028d5f096d3bb76672fa2475080f31L131
Adding this extra check to the query does not solve the issue addressed on the test anyway. Still I'm not sure about the flows :/

I want to point out also that this DB could have been upgraded to 13.0 before the original PR was merged (I've not checked the dates precisely), therefore they may be reminiscences of previous flows. I also confirm that this is reproducible also in 14.0

If this patch is found to be safe, even if not covered by a valid flow, it may be helpful for upgrades.

Thanks a lot for your input and the details on the flow in the test.

Reply:
My previous comment was too much talk and little details. Below the details of the failing upgrade that could benefit from this PR.
Before upgrade, on v13:
```
23410_afu=> select id,name,key,inherit_id,website_id,mode,type from ir_ui_view where key in ('portal.frontend_layout','website.layout','website.template_header_default','website.template_header_default','website_s
 ale.template_header_default','web.frontend_layout','web.layout','website.footer_custom') order by key,id
+------+----------------------+------------------------+--------------+--------------+-----------+--------+
| id   | name                 | key                    | inherit_id   | website_id   | mode      | type   |
|------+----------------------+------------------------+--------------+--------------+-----------+--------|
| 371  | Main Frontend Layout | portal.frontend_layout | 6163         | 1            | extension | qweb   |
| 3659 | Main Frontend Layout | portal.frontend_layout | 6163         | <null>       | extension | qweb   |
| 6163 | Frontend Layout      | web.frontend_layout    | 171          | 1            | primary   | qweb   |
| 171  | Web layout           | web.layout             | <null>       | <null>       | primary   | qweb   |
| 478  | Footer               | website.footer_custom  | 471          | 1            | extension | qweb   |
| 3681 | Footer               | website.footer_custom  | 3662         | 1            | extension | qweb   |
| 8377 | Footer               | website.footer_custom  | 471          | <null>       | extension | qweb   |
| 471  | Main layout          | website.layout         | 371          | <null>       | extension | qweb   |
| 3662 | Main layout          | website.layout         | 3659         | <null>       | extension | qweb   |
+------+----------------------+------------------------+--------------+--------------+-----------+--------+
```
Which translates to (node is `key website_id`):
```
            171 NULL
               |
             6163 1
            /      \
    3659 NULL =key= 371 1
        /              \
  3662 NULL   =key=   471 NULL
      /             /         \
3681 1 =key= 8377 NULL =key= 478 1
```
During upgrade to 14.0, right before failing with:
```
psycopg2.errors.ForeignKeyViolation: insert or update on table "ir_ui_view" violates foreign key constraint "ir_ui_view_inherit_id_fkey"
DETAIL:  Key (inherit_id)=(9867) is not present in table "ir_ui_view".
```
We have
```
23410_afu_14.0=> select id,name,key,inherit_id,website_id,mode,type from ir_ui_view where key in ('portal.frontend_layout','website.layout','website.template_header_default','website.template_header_default','webs
 ite_sale.template_header_default','web.frontend_layout','web.layout','website.footer_custom') order by key,id
+-------+-------------------------+--------------------------------------+--------------+--------------+-----------+--------+
| id    | name                    | key                                  | inherit_id   | website_id   | mode      | type   |
|-------+-------------------------+--------------------------------------+--------------+--------------+-----------+--------|
| 371   | Main Frontend Layout    | portal.frontend_layout               | 6163         | 1            | extension | qweb   |
| 3659  | Main Frontend Layout    | portal.frontend_layout               | 6163         | <null>       | extension | qweb   |
| 6163  | Frontend Layout         | web.frontend_layout                  | 171          | 1            | primary   | qweb   |
| 171   | Web layout              | web.layout                           | <null>       | <null>       | primary   | qweb   |
| 478   | Default                 | website.footer_custom                | 471          | 1            | extension | qweb   |
| 3681  | Default                 | website.footer_custom                | 3662         | 1            | extension | qweb   |
| 8377  | Default                 | website.footer_custom                | 471          | <null>       | extension | qweb   |
| 471   | Main layout             | website.layout                       | 371          | <null>       | extension | qweb   |
| 3662  | Main layout             | website.layout                       | 3659         | <null>       | extension | qweb   |
| 9556  | Template Header Default | website.template_header_default      | 471          | <null>       | extension | qweb   |
| 9867  | Template Header Default | website.template_header_default      | 471          | 1            | extension | qweb   |
| 10055 | template_header_default | website_sale.template_header_default | 9556         | <null>       | extension | qweb   |
+-------+-------------------------+--------------------------------------+--------------+--------------+-----------+--------+
```
In tree form:
```
            171 NULL
               |
             6163 1
            /      \
    3659 NULL =key= 371 1
        /              \
  3662 NULL   =key=   471 NULL -----+---------------+
      /             /         \      \               \
3681 1 =key= 8377 NULL =key= 478 1    \               \ 
                                      9556 null =key= 9867 1
                                             |
                                       10055 null 
```
Below the content of the `result`[1] dict (highlighted the relevant views)
```
{471: [[371], [1]],       ***
 3051: [[6163], [1]],
 3662: [[371], [1]],      
 6800: [[3666], [1]],
 6801: [[3666], [1]],
 6814: [[3085], [1]],
 6815: [[3085], [1]],
 6816: [[3089], [1]],
 6825: [[3085], [1]],
 9423: [[371], [1]],
 9451: [[6163], [1]],
 9927: [[3666], [1]],
 10006: [[6163], [1]],
 10055: [[9867], [1]],     ***
 10056: [[9871], [1]],
 10068: [[3085], [1]],
 10070: [[3085], [1]]}
```
471 needs to be processed since it doesn't have a specific counterpart. While doing so, since 9867 is one of its specific children, 9867 gets recreated and unlinked[2]. That's fine.
Now, when 10055 is processed, its specific counterpart's parent should be 9867 (as computed in the SQL query, and reflected in `result`). The issue is that when the new view is created, 9867 is already gone. Thus we get the traceback.

Also, `len(processed_modules)`[3] is 238, thus it won't be surprising that many views have been touched (created, removed, changed).

[1] https://github.com/odoo/odoo/blob/666955cb630640384c5ca1f0e13e13b5253c3c58/addons/website/models/ir_ui_view.py#L130-L132
[2] https://github.com/odoo/odoo/blob/666955cb630640384c5ca1f0e13e13b5253c3c58/addons/website/models/ir_ui_view.py#L88
[3] https://github.com/odoo/odoo/blob/666955cb630640384c5ca1f0e13e13b5253c3c58/addons/website/models/ir_ui_view.py#L105